### PR TITLE
feat: Font list cache

### PR DESCRIPTION
**Pull request by xfqz86**
refactor: Extract a common font settings component

### CODE REVIEW

File: src/defaults.ts [hunk 1]
```diff
@@ -45,7 +45,7 @@ export const OBJECT_DEFAULTS_VECTOR = {
 export const OBJECT_DEFAULTS_TEXT = {
   ...OBJECT_DEFAULTS,
   fill: "black",
-  fontFamily: "Noto Sans Variable",
+  fontFamily: "Verdana",
```
Comment (MultiMote) [inline src/defaults.ts:48]: Why? Noto Sans is bundled with niimblue.
https://github.com/MultiMote/niimblue/blob/20c1e6c012cfd3032b7fce193e672a368ad0caf1/index.html#L16-L20

File: src/lib/BarcodeParamsControls.svelte [hunk 2]
```diff
@@ -53,20 +54,21 @@
...
+      family={selectedBarcode.fontFamily}
+      size={selectedBarcode.fontSize}
+      on:change={(e) => {
+        const {
+          newValue: { family, size },
+        } = e.detail;
+        selectedBarcode?.set({
+          fontFamily: family,
+          fontSize: size,
+        });
```
Comment (MultiMote) [inline src/lib/BarcodeParamsControls.svelte:68]: Action is too long. It should be extracted to function.

File: src/lib/TextParamsControls.svelte
Comment (MultiMote) [inline src/lib/TextParamsControls.svelte:1]: getFonts is unused now

File: src/lib/basic/FontParams.svelte [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+  export let size: number = 0;
+  export let sizeMin: number = 1;
+  export let sizeMax: number = 999;
+  export let sizeStep: number = 2;
+
+  const dispatch = createEventDispatcher<{
+    change: { newValue: ChangeValue };
+  }>();
+
+  let familys: Record<string, string> = {};
```
Comment (MultiMote) [inline src/lib/basic/FontParams.svelte:20]: "families"

File: src/lib/basic/FontParams.svelte [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+  const fontQuerySupported = typeof queryLocalFonts !== "undefined";
+  async function refreshFonts() {
+    if (fontQuerySupported) {
+      window.fontCache = await queryLocalFonts();
```
Comment (MultiMote) [inline src/lib/basic/FontParams.svelte:25]: Use svelte stores for global data

https://github.com/MultiMote/niimblue/blob/20c1e6c012cfd3032b7fce193e672a368ad0caf1/src/stores.ts#L20

**Review by MultiMote — CHANGES_REQUESTED** [review]

**Comment by xfqz86** [comment]
fixed

File: src/lib/basic/FontParams.svelte [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+<script lang="ts">
+  import { createEventDispatcher, onDestroy, onMount } from "svelte";
+  import { tr } from "../../utils/i18n";
+  import MdIcon from "./MdIcon.svelte";
+  import { fontCache } from "../../stores";
+  import type { Unsubscriber } from "svelte/motion";
+
+  export let hasFamily: boolean = true;
+  export let hasSize: boolean = true;
```
Comment (MultiMote) [inline src/lib/basic/FontParams.svelte:14]: It has no use in this project. Should be removed.

File: src/lib/basic/FontParams.svelte [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+      },
+    });
+  }
+
+  let unsubscriber: Unsubscriber;
+  onMount(() => {
+    resetFamilyAndStyle();
+    unsubscriber = fontCache.subscribe(resetFamilyAndStyle);
+  });
+  onDestroy(() => unsubscriber());
```
Comment (MultiMote) [inline src/lib/basic/FontParams.svelte:74]: entire block can be replaced with `$: resetFamilyAndStyle($fontCache);`

File: src/lib/basic/FontParams.svelte
Comment (MultiMote) [inline src/lib/basic/FontParams.svelte:1]: What is the logic of component splitting? Why text size goes to a new component, but line height, bold are not?

File: src/lib/TextParamsControls.svelte
Comment (MultiMote) [inline src/lib/TextParamsControls.svelte:1]: fontQuerySupported, fontFamilies are unused

File: src/stores.ts [hunk 7]
```diff
@@ -34,6 +34,7 @@ export const automation = readable<AutomationProps | undefined>(
     return undefined;
   })()
 );
+export const fontCache = writable<ReadonlyArray<FontData> | undefined>(undefined);
```
Comment (MultiMote) [inline src/stores.ts:37]: (undefined) → ()

File: src/lib/basic/FontParams.svelte [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+{#if hasFamily}
+  <div class="input-group flex-nowrap input-group-sm font-family">
+    <span class="input-group-text" title={$tr("params.text.font_family")}>
+      <MdIcon icon="text_format" />
+    </span>
+    {#if fontQuerySupported}
+      <select class="form-select" value={family} on:change={(e) => emitChange({ family: e.currentTarget.value })}>
+        {#each Object.keys(families).sort() as family}
+          <option value={family} style="font-family: {family}">{families[family]}</option>
```
Comment (MultiMote) [inline src/lib/basic/FontParams.svelte:85]: You are using wrong font names. Styles are not applied

image

**Review by MultiMote — CHANGES_REQUESTED** [review]

**Comment by MultiMote** [comment]
Implemented in https://github.com/MultiMote/niimblue/commit/9cab3d2c3b2854dc39c5c014a719042f6bc58354

P.S. Please don't move all business logic to template segment. I was inattentive and missed that in past commits.

**Pull request closed without merging**